File: src/csharp/core/src/AnimationSelection.cs
using System.Collections.Generic;

namespace DiceBear.Internal
{
    /// <summary>
    /// The interpreted <c>animation</c> option: off, all timelines, or a
    /// selection of timeline names.
    /// </summary>
    /// <remarks>
    /// Built from the raw memoized option value, so the resolved-options
    /// snapshot keeps the user's shape (a boolean, or the name list in the
    /// given order) while consumers work with these accessors.
    /// </remarks>
    internal readonly struct AnimationSelection
    {
        private readonly bool _all;
        private readonly IReadOnlyList<string>? _names;

        private AnimationSelection(bool all, IReadOnlyList<string>? names)
        {
            _all = all;
            _names = names;
        }

        internal static AnimationSelection From(object? raw)
        {
            switch (raw)
            {
                case bool flag:
                    return new AnimationSelection(flag, null);
                case IReadOnlyList<string> names:
                    return new AnimationSelection(false, names);
                default:
                    return new AnimationSelection(false, null);
            }
        }

        /// <summary>Whether no timeline plays at all.</summary>
        internal bool Off => !_all && (_names is null || _names.Count == 0);

        /// <summary>
        /// Whether a timeline carrying the given name plays. <c>true</c>
        /// plays every timeline. A name selection plays only named timelines
        /// carrying one of the selected names.
        /// </summary>
        internal bool Matches(string? name)
        {
            if (_all)
            {
                return true;
            }

            if (name is null || _names is null)
            {
                return false;
            }

            foreach (var candidate in _names)
            {
                if (candidate == name)
                {
                    return true;
                }
            }

            return false;
        }

        /// <summary>
        /// The selected names in user order, or <see langword="null"/> for
        /// the boolean forms. Consumed by the animation class namespace hash.
        /// </summary>
        internal IReadOnlyList<string>? Names => _names;
    }
}

File: src/dart/core/lib/src/animation_selection.dart
/// The interpreted `animation` option: off, all timelines, or a selection of
/// timeline names.
library;

/// The interpreted `animation` option: off, all timelines, or a selection of
/// timeline names.
///
/// Built from the raw memoized option value, so the resolved-options snapshot
/// keeps the user's shape (a boolean, or the name list in the given order)
/// while consumers work with these accessors.
class AnimationSelection {
  final bool _all;
  final List<String>? _names;

  const AnimationSelection._(this._all, this._names);

  factory AnimationSelection.from(Object? raw) {
    if (raw is bool) {
      return AnimationSelection._(raw, null);
    }

    if (raw is List<String>) {
      return AnimationSelection._(false, raw);
    }

    return const AnimationSelection._(false, null);
  }

  /// Whether no timeline plays at all.
  bool get off {
    final names = _names;

    return !_all && (names == null || names.isEmpty);
  }

  /// Whether a timeline carrying the given [name] plays. `true` plays every
  /// timeline. A name selection plays only named timelines carrying one of
  /// the selected names.
  bool matches(String? name) {
    if (_all) {
      return true;
    }

    final names = _names;

    if (name == null || names == null) {
      return false;
    }

    return names.contains(name);
  }

  /// The selected names in user order, or `null` for the boolean forms.
  /// Consumed by the animation class namespace hash.
  List<String>? get names => _names;
}
